### PR TITLE
Add supporting regexes to packages API endpoint

### PR DESCRIPTION
**Pull request by blayson**

### CODE REVIEW

**Comment by psegedy** [comment]
In jenkins I see some differences from previous behavior:
1. packages that are not in DB were previously printed in package list, it's not a bug, just tell me how do you want to have it so I can update tests appropriately.
```
➜  iqe-vulnerability-plugin git:(psegedy/no-sm-upload) ✗ curl http://localhost:8080/api/v1/packages/non-existing                     
{"package_list": {}}%             
                                                                                                                                                                                                             
➜  iqe-vulnerability-plugin git:(psegedy/no-sm-upload) ✗ curl https://webapp-vmaas-ci.5a9f.insights-dev.openshiftapps.com/api/v1/packages/non-existing                     
{"package_list": {"non-existing": {}}}%    
```
2. Package with epoch=0 should return same result when epoch is specified or not.
```
➜  iqe-vulnerability-plugin git:(psegedy/no-sm-upload) ✗ curl http://localhost:8080/api/v1/packages/bash-0:4.2.46-20.el7_2.x86_64                                      
{"package_list": {}}%                                                                                                                                                                                                                          
➜  iqe-vulnerability-plugin git:(psegedy/no-sm-upload) ✗ curl http://localhost:8080/api/v1/packages/bash-4.2.46-20.el7_2.x86_64 
{"package_list": {"bash-4.2.46-20.el7_2.x86_64": {"summary": "The GNU Bourne Again shell", "description": "The GNU Bourne Again shell (Bash) is a shell or command language\ninterpreter that is compatible with the Bourne shell (sh). Bash\nincorporates useful features from the Korn shell (ksh) and the C shell\n(csh). Most sh scripts can be run by bash without modification.", "source_package": "bash-4.2.46-20.el7_2.src", "repositories": [{"label": "rhel-7-server-rpms", "name": "Red Hat Enterprise Linux 7 Server (RPMs)", "basearch": "x86_64", "releasever": "7Server"}, {"label": "rhel-7-workstation-rpms", "name": "Red Hat Enterprise Linux 7 Workstation (RPMs)", "basearch": "x86_64", "releasever": "7Workstation"}], "package_list": []}}}
```

**Comment by jdobes** [comment]
Also getting source packages doesn't work:

```
curl -X GET http://localhost:8080/api/v1/packages/rsyslog7-7.4.10-3.el6_6.src
{"package_list": {}}
```

on CI:

```
curl -X GET https://webapp-vmaas-ci.5a9f.insights-dev.openshiftapps.com/api/v1/packages/rsyslog7-7.4.10-3.el6_6.src
{"package_list": {"rsyslog7-7.4.10-3.el6_6.src": {"summary": null, "description": null, "source_package": null, "repositories": [], "package_list": ["rsyslog7-pgsql-7.4.10-3.el6_6.i686", "rsyslog7-pgsql-7.4.10-3.el6_6.ppc64", "rsyslog7-pgsql-7.4.10-3.el6_6.x86_64", "rsyslog7-pgsql-7.4.10-3.el6_6.s390x", "rsyslog7-mysql-7.4.10-3.el6_6.i686", "rsyslog7-mysql-7.4.10-3.el6_6.ppc64", "rsyslog7-mysql-7.4.10-3.el6_6.x86_64", "rsyslog7-mysql-7.4.10-3.el6_6.s390x", "rsyslog7-elasticsearch-7.4.10-3.el6_6.s390x", "rsyslog7-elasticsearch-7.4.10-3.el6_6.ppc64", "rsyslog7-elasticsearch-7.4.10-3.el6_6.i686", "rsyslog7-elasticsearch-7.4.10-3.el6_6.x86_64", "rsyslog7-7.4.10-3.el6_6.i686", "rsyslog7-7.4.10-3.el6_6.ppc64", "rsyslog7-7.4.10-3.el6_6.x86_64", "rsyslog7-7.4.10-3.el6_6.s390x", "rsyslog7-snmp-7.4.10-3.el6_6.s390x", "rsyslog7-snmp-7.4.10-3.el6_6.ppc64", "rsyslog7-snmp-7.4.10-3.el6_6.i686", "rsyslog7-snmp-7.4.10-3.el6_6.x86_64", "rsyslog7-debuginfo-7.4.10-3.el6_6.s390x", "rsyslog7-debuginfo-7.4.10-3.el6_6.ppc64", "rsyslog7-debuginfo-7.4.10-3.el6_6.i686", "rsyslog7-debuginfo-7.4.10-3.el6_6.x86_64", "rsyslog7-gnutls-7.4.10-3.el6_6.i686", "rsyslog7-gnutls-7.4.10-3.el6_6.ppc64", "rsyslog7-gnutls-7.4.10-3.el6_6.x86_64", "rsyslog7-gnutls-7.4.10-3.el6_6.s390x", "rsyslog7-gssapi-7.4.10-3.el6_6.i686", "rsyslog7-gssapi-7.4.10-3.el6_6.ppc64", "rsyslog7-gssapi-7.4.10-3.el6_6.x86_64", "rsyslog7-gssapi-7.4.10-3.el6_6.s390x", "rsyslog7-relp-7.4.10-3.el6_6.i686", "rsyslog7-relp-7.4.10-3.el6_6.ppc64", "rsyslog7-relp-7.4.10-3.el6_6.x86_64", "rsyslog7-relp-7.4.10-3.el6_6.s390x"]}}}
```

File: webapp/packages.py [hunk 1]
```diff
@@ -23,6 +24,23 @@ class PackagesAPI:
...
+        if not regex.startswith('^'):
+            regex = '^' + regex
+
+        if not regex.endswith('$'):
+            regex = regex + '$'
+
+        filtered_pkgs = []
+        for (name_id, evr_id, arch_id) in self.cache.nevra2pkgid:
+            pkg_id = self.cache.nevra2pkgid[(name_id, evr_id, arch_id)]
+            package_list = self._get_built_binary_packages(pkg_id)
```
Comment (jdobes) [inline webapp/packages.py]: this is probably overly complicated and doesn't work for source packages

would something like this work?
```
for (name_id, evr_id, arch_id) in self.cache.nevra2pkgid:
    # get name, epoch, version, release, arch
    # ...
    variants = ["%s-%s:%s-%s.%s" % (name, epoch, version, release, arch)]
    if epoch == 0:
        variants.append("%s-%s-%s.%s" % (name, version, release, arch))
    for item in variants:
        if re.match(regex, item):
            filtered_pkgs.append(item)
```

Comment (blayson) [reply webapp/packages.py]: I found a little bit another way how to do it
```
        regex = packages[0]
        name, epoch, ver, rel, arch = utils.split_packagename(regex)
        if name and epoch and ver and rel and arch:
            # If there is no regular expression return all packages what we get in api request
            return packages
```
Looks like it works fine

File: webapp/packages.py [hunk 2]
```diff
@@ -53,6 +76,10 @@ def process_list(self, api_version, data): # pylint: disable=unused-argument
         if not packages:
             return packagelist
 
+        if len(packages) == 1:
+            # treat single-label like a regex, get all matching names
+            packages = self.find_packages_by_regex(packages[0])
```
Comment (jdobes) [inline webapp/packages.py]: this doesn't work, on next line is `for pkg in packages:` and it produces something like this:

```
curl -X GET http://localhost:8080/api/v1/packages/pam_ssh_agent_auth-0.9-70.el6.x86_64
{"package_list": {"p": {}, "a": {}, "m": {}, "_": {}, "s": {}, "h": {}, "g": {}, "e": {}, "n": {}, "t": {}, "u": {}, "-": {}, "0": {}, ".": {}, "9": {}, "7": {}, "l": {}, "6": {}, "x": {}, "8": {}, "4": {}}}
```

File: webapp/test/test_packages.py [hunk 3]
```diff
@@ -10,6 +10,7 @@
 from packages import PackagesAPI
 
 PKG = "my-pkg-1.1.0-1.el8.i686"
+PKG_EPOCH = "my-pkg-0:1.1.0-1.el8.i686"
```
Comment (jdobes) [inline webapp/test/test_packages.py]: this variable isn't used

**Comment by blayson** [comment]
I made some changes to a condition, but I don't know if it is a good approach

**Comment by jdobes** [comment]
Closing this PR, please re-open if you find better solution.

**Pull request closed without merging**